File: app/src/main/java/com/nononsenseapps/feeder/ui/ReaderFragment.java
/*
 * Copyright (c) 2015 Jonas Kalderstam.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.nononsenseapps.feeder.ui;


import android.content.ActivityNotFoundException;
import android.content.Intent;
import android.database.Cursor;
import android.graphics.Point;
import android.net.Uri;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.CursorLoader;
import android.support.v4.content.Loader;
import android.support.v4.view.MenuItemCompat;
import android.support.v7.widget.ShareActionProvider;
import android.text.Spanned;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;
import com.nononsenseapps.feeder.R;
import com.nononsenseapps.feeder.db.FeedItemSQL;
import com.nononsenseapps.feeder.db.FeedItemSQLKt;
import com.nononsenseapps.feeder.ui.text.HtmlConverter;
import com.nononsenseapps.feeder.ui.text.ImageTextLoader;
import com.nononsenseapps.feeder.util.BundleExtensionsKt;
import com.nononsenseapps.feeder.util.ContentResolverExtensionsKt;
import com.nononsenseapps.feeder.util.FileLog;
import com.nononsenseapps.feeder.util.PrefUtils;
import com.nononsenseapps.feeder.util.TabletUtils;
import com.nononsenseapps.feeder.views.ObservableScrollView;
import org.joda.time.DateTimeZone;
import org.joda.time.format.DateTimeFormat;
import org.joda.time.format.DateTimeFormatter;

import java.util.Locale;

import static com.nononsenseapps.feeder.db.FeedItemSQLKt.FEED_ITEM_FIELDS;
import static com.nononsenseapps.feeder.db.FeedSQLKt.COL_ID;
import static com.nononsenseapps.feeder.db.UriKt.URI_FEEDITEMS;

/**
 * A simple {@link Fragment} subclass.
 * Use the {@link ReaderFragment#newInstance} factory method to
 * create an instance of this fragment.
 */
public class ReaderFragment extends Fragment
        implements LoaderManager.LoaderCallbacks {
    public static final String ARG_TITLE = "title";
    public static final String ARG_DESCRIPTION = "body";
    public static final String ARG_LINK = "link";
    public static final String ARG_ENCLOSURE = "enclosure";
    public static final String ARG_IMAGEURL = "imageurl";
    public static final String ARG_ID = "dbid";
    public static final String ARG_FEEDTITLE = "feedtitle";
    public static final String ARG_AUTHOR = "author";
    public static final String ARG_DATE = "date";
    // TODO Change
    static final DateTimeFormatter dateTimeFormat =
            DateTimeFormat.mediumDate().withLocale(Locale.getDefault());
    private static final int TEXT_LOADER = 1;
    private static final int ITEM_LOADER = 2;
    // TODO database id
    private long _id = -1;
    // All content contained in RssItem
    private FeedItemSQL mRssItem;
    private TextView mBodyTextView;
    private ObservableScrollView mScrollView;
    private TextView mTitleTextView;


    public ReaderFragment() {
    }

    /**
     * Use this factory method to create a new instance of
     * this fragment using the provided parameters.
     *
     * @param rssItem the Item to open in the reader
     * @return A new instance of fragment ReaderFragment.
     */
    public static ReaderFragment newInstance(FeedItemSQL rssItem) {
        ReaderFragment fragment = new ReaderFragment();
        // Save some time on load
        fragment.mRssItem = rssItem;
        fragment._id = rssItem.getId();

        fragment.setArguments(RssItemToBundle(rssItem, null));
        return fragment;
    }

    /**
     * Convert an RssItem into a Bundle for use with Fragment Arguments
     *
     * @param rssItem to convert
     * @param bundle  may be null
     * @return bundle of rssItem plus id
     */
    public static Bundle RssItemToBundle(FeedItemSQL rssItem,
                                         Bundle bundle) {
        if (bundle == null) {
            bundle = new Bundle();
        }
        bundle.putLong(ARG_ID, rssItem.getId());
        bundle.putString(ARG_TITLE, rssItem.getTitle());
        bundle.putString(ARG_LINK, rssItem.getLink());
        bundle.putString(ARG_ENCLOSURE, rssItem.getEnclosurelink());
        bundle.putString(ARG_IMAGEURL, rssItem.getImageurl());
        bundle.putString(ARG_FEEDTITLE, rssItem.getFeedtitle());
        bundle.putString(ARG_AUTHOR, rssItem.getAuthor());
        bundle.putString(ARG_DATE, rssItem.getPubDateString());
        return bundle;
    }

    public static FeedItemSQL RssItemFromBundle(Bundle bundle) {
        return BundleExtensionsKt.asFeedItem(bundle);
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (savedInstanceState != null) {
            _id = savedInstanceState.getLong(ARG_ID);
            mRssItem = RssItemFromBundle(savedInstanceState);

        } else if (mRssItem == null) {
            // Construct from arguments
            _id = getArguments().getLong(ARG_ID, -1);
            mRssItem = RssItemFromBundle(getArguments());
        }

        if (_id > 0) {
            ContentResolverExtensionsKt.markItemAsRead(getActivity().getContentResolver(), _id, true);
            getLoaderManager().restartLoader(ITEM_LOADER, new Bundle(), this);
        }

        setHasOptionsMenu(true);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        final int theLayout;
        if (TabletUtils.isTablet(getActivity())) {
            theLayout = R.layout.fragment_reader_tablet;
        } else {
            theLayout = R.layout.fragment_reader;
        }
        View rootView = inflater.inflate(theLayout, container, false);

        mScrollView =
                (ObservableScrollView) rootView.findViewById(R.id.scroll_view);
        mTitleTextView = (TextView) rootView.findViewById(R.id.story_title);
        mBodyTextView = (TextView) rootView.findViewById(R.id.story_body);
        TextView mAuthorTextView = (TextView) rootView.findViewById(R.id.story_author);
        TextView mFeedTitleTextView = (TextView) rootView.findViewById(R.id
                .story_feedtitle);

        setViewTitle();

        mFeedTitleTextView.setText(mRssItem.getFeedtitle());

        if (mRssItem.getAuthor() == null && mRssItem.getPubDate() != null) {
            mAuthorTextView.setText(getString(R.string.on_date,
                    mRssItem.getPubDate().withZone(DateTimeZone.getDefault())
                            .toString(dateTimeFormat)));
        } else if (mRssItem.getPubDate() != null) {
            mAuthorTextView.setText(getString(R.string.by_author_on_date,
                    mRssItem.getAuthor(),
                    mRssItem.getPubDate().withZone(DateTimeZone.getDefault())
                            .toString(dateTimeFormat)));
        } else {
            mAuthorTextView.setVisibility(View.GONE);
        }

        setViewBody();

        return rootView;
    }

    private void setViewTitle() {
        if (mRssItem.getTitle().isEmpty()) {
            mTitleTextView.setText(mRssItem.getPlaintitle());
        } else {
            mTitleTextView.setText(HtmlConverter.toSpannedWithNoImages(mRssItem.getTitle(), getActivity()));
        }
    }

    private void setViewBody() {
        if (!mRssItem.getDescription().isEmpty()) {
            // Set without images as a place holder
            mBodyTextView.setText(HtmlConverter.toSpannedWithNoImages(mRssItem.getDescription(), getActivity()));

            // Load images in text
            getLoaderManager().restartLoader(TEXT_LOADER, new Bundle(), this);
        }
    }

    @Override
    public void onActivityCreated(Bundle bundle) {
        super.onActivityCreated(bundle);
        // TODO set for scrollview
        ((BaseActivity) getActivity()).enableActionBarAutoHide(mScrollView);
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        RssItemToBundle(mRssItem, outState);
        super.onSaveInstanceState(outState);
    }

    @Override
    public void onDetach() {
        getLoaderManager().destroyLoader(TEXT_LOADER);
        super.onDetach();
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.reader, menu);

        // Locate MenuItem with ShareActionProvider
        MenuItem shareItem = menu.findItem(R.id.action_share);

        // Fetch and store ShareActionProvider
        ShareActionProvider shareActionProvider =
                (ShareActionProvider) MenuItemCompat.getActionProvider(shareItem);

        // Set intent
        Intent shareIntent = new Intent(Intent.ACTION_SEND);
        shareIntent.setType("text/plain");
        shareIntent.putExtra(Intent.EXTRA_TEXT, mRssItem.getLink());
        shareActionProvider.setShareIntent(shareIntent);

        // Show/Hide enclosure
        menu.findItem(R.id.action_open_enclosure).setVisible(mRssItem.getEnclosurelink() != null);
        // Add filename to tooltip
        if (mRssItem.getEnclosurelink() != null) {
            String filename = mRssItem.getEnclosureFilename();
            if (filename != null) {
                menu.findItem(R.id.action_open_enclosure).setTitle(filename);
            }
        }

        // Don't forget super call here
        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem menuItem) {
        Uri uri;
        switch (menuItem.getItemId()) {
            case R.id.action_open_in_browser:
                uri = Uri.parse(mRssItem.getLink());
                if (uri.isRelative()) {
                    Toast.makeText(getActivity(), "Sorry, can't handle relative links yet.", Toast.LENGTH_SHORT).show();
                }
                // Open in browser
                try {
                    startActivity(new Intent(Intent.ACTION_VIEW, uri));
                } catch (ActivityNotFoundException e) {
                    Toast.makeText(getActivity(), "Couldn't find an activity to open that link with", Toast.LENGTH_SHORT).show();
                    FileLog.singleton.getInstance(getActivity()).d("No such activity: " + e);
                }
                return true;
            case R.id.action_open_enclosure:
                uri = Uri.parse(mRssItem.getEnclosurelink());
                if (uri.isRelative()) {
                    Toast.makeText(getActivity(), R.string.no_activity_for_link, Toast.LENGTH_SHORT).show();
                }
                // Open enclosure link
                try {
                    startActivity(new Intent(Intent.ACTION_VIEW, uri));
                } catch (ActivityNotFoundException e) {
                    Toast.makeText(getActivity(), R.string.no_activity_for_link, Toast.LENGTH_SHORT).show();
                    FileLog.singleton.getInstance(getActivity()).d("No such activity: " + e);
                }
                return true;
            default:
                return super.onOptionsItemSelected(menuItem);
        }
    }

    @Override
    public Loader onCreateLoader(final int id, final Bundle args) {
        if (id == ITEM_LOADER) {
            return new CursorLoader(getContext(), URI_FEEDITEMS,
                    FEED_ITEM_FIELDS,
                    COL_ID + " IS ?",
                    new String[]{Long.toString(mRssItem.getId())},
                    null);
        } else if (id == TEXT_LOADER) {
            Point size = new Point();
            getActivity().getWindowManager().getDefaultDisplay().getSize(size);
            // Using twice window height since we do scroll vertically
            if (TabletUtils.isTablet(getActivity())) {
                // Tablet has fixed width
                return new ImageTextLoader(getActivity(), mRssItem.getDescription(),
                        new Point(Math.round(getResources().getDimension(R.dimen.reader_tablet_width)), 2 * size.y), PrefUtils.shouldLoadImages(getActivity()));
            } else {
                // Base it on window size
                return new ImageTextLoader(getActivity(), mRssItem.getDescription(),
                        new Point(size.x - 2 * Math.round(getResources().getDimension(R.dimen.keyline_1)), 2 * size.y), PrefUtils.shouldLoadImages(getActivity()));
            }
        }
        return null;
    }


    @Override
    public void onLoadFinished(final Loader loader,
                               final Object data) {
        if (loader.getId() == ITEM_LOADER) {
            Cursor cursor = (Cursor) data;
            if (cursor != null) {
                if (cursor.moveToFirst()) {
                    mRssItem = FeedItemSQLKt.asFeedItem((Cursor) data);
                    setViewTitle();
                    setViewBody();
                }
                cursor.close();
            }
            loader.reset();
        } else if (loader.getId() == TEXT_LOADER) {
            if (data != null) {
                mBodyTextView.setText((Spanned) data);
            }
            loader.reset();
        }
    }

    @Override
    public void onLoaderReset(final Loader loader) {
        // nothing really
    }
}

File: app/src/main/java/com/nononsenseapps/feeder/ui/ReaderFragment.kt
package com.nononsenseapps.feeder.ui

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.text.Spanned
import android.util.Log
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.core.text.BidiFormatter
import androidx.fragment.app.Fragment
import androidx.lifecycle.LiveData
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.observe
import androidx.navigation.fragment.findNavController
import com.nononsenseapps.feeder.R
import com.nononsenseapps.feeder.base.KodeinAwareFragment
import com.nononsenseapps.feeder.db.room.FeedItemWithFeed
import com.nononsenseapps.feeder.db.room.ID_UNSET
import com.nononsenseapps.feeder.model.FeedItemViewModel
import com.nononsenseapps.feeder.model.TextToSpeechViewModel
import com.nononsenseapps.feeder.model.TextOptions
import com.nononsenseapps.feeder.model.cancelNotification
import com.nononsenseapps.feeder.model.maxImageSize
import com.nononsenseapps.feeder.util.PREF_VAL_OPEN_WITH_CUSTOM_TAB
import com.nononsenseapps.feeder.util.PREF_VAL_OPEN_WITH_WEBVIEW
import com.nononsenseapps.feeder.util.Prefs
import com.nononsenseapps.feeder.util.TabletUtils
import com.nononsenseapps.feeder.util.bundle
import com.nononsenseapps.feeder.util.openLinkInBrowser
import com.nononsenseapps.feeder.util.openLinkInCustomTab
import kotlinx.coroutines.FlowPreview
import kotlinx.coroutines.launch
import org.kodein.di.Kodein
import org.kodein.di.android.x.closestKodein
import org.kodein.di.generic.instance
import org.threeten.bp.format.DateTimeFormatter
import org.threeten.bp.format.FormatStyle
import java.util.Locale

const val ARG_TITLE = "title"
const val ARG_CUSTOMTITLE = "customtitle"
const val ARG_LINK = "link"
const val ARG_ENCLOSURE = "enclosure"
const val ARG_IMAGEURL = "imageUrl"
const val ARG_ID = "dbid"
const val ARG_AUTHOR = "author"
const val ARG_DATE = "date"

@FlowPreview
class ReaderFragment : KodeinAwareFragment() {
    private val dateTimeFormat =
        DateTimeFormatter.ofLocalizedDateTime(FormatStyle.FULL, FormatStyle.SHORT)
            .withLocale(Locale.getDefault())

    private var _id: Long = ID_UNSET

    private lateinit var titleTextView: TextView
    private lateinit var bodyTextView: TextView
    private lateinit var authorTextView: TextView
    private lateinit var feedTitleTextView: TextView

    private val viewModel: FeedItemViewModel by instance(arg = this)

    private val prefs: Prefs by instance()
    private val warmer: CustomTabsWarmer by instance()

    private var liveText: LiveData<Spanned>? = null

    // For menu things
    private var rssItemMenu: FeedItemWithFeed? = null

    init {
        lifecycleScope.launchWhenStarted {
            try {
                if (prefs.shouldPreloadCustomTab) {
                    val rssItem = viewModel.getItem(_id)
                    warmer.preLoad {
                        rssItem.link?.let { Uri.parse(it) }
                    }
                }
            } catch (e: Exception) {
                // Don't let this crash
                Log.e("FeederReaderFragment", "Couldn't preload $_id", e)
            }
        }
    }

    private fun getTextOptions(): TextOptions =
        TextOptions(
            itemId = _id,
            maxImageSize = requireActivity().maxImageSize(),
            nightMode = prefs.isNightMode
        )

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        // Fragment's views have their own coroutine scope
        viewLifecycleOwner.lifecycleScope.launchWhenCreated {
            // It may have been manually overridden by fetching full text
            if (liveText == null) {
                liveText = viewModel.getLiveTextMaybeFull(
                    getTextOptions(),
                    urlClickListener()
                )
            }

            // Fragment's views have their own lifecycle
            liveText?.observe(viewLifecycleOwner) {
                bodyTextView.text = it
            }
        }
        viewModel.getLiveItem(_id).observe(viewLifecycleOwner) { rssItem ->
            if (rssItem == null) {
                return@observe
            }
            rssItemMenu = rssItem

            titleTextView.text = rssItem.plainTitle

            rssItem.feedId?.let { feedId ->
                feedTitleTextView.setOnClickListener {
                    findNavController()
                        .navigate(
                            R.id.action_readerFragment_to_feedFragment,
                            bundle {
                                putLong(ARG_FEED_ID, feedId)
                            }
                        )
                }
            }

            feedTitleTextView.text = rssItem.feedDisplayTitle

            rssItem.pubDate.let { pubDate ->
                rssItem.author.let { author ->
                    when {
                        author == null && pubDate != null ->
                            authorTextView.text = getString(
                                R.string.on_date,
                                pubDate.format(dateTimeFormat)
                            )
                        author != null && pubDate != null ->
                            authorTextView.text = getString(
                                R.string.by_author_on_date,
                                // Must wrap author in unicode marks to ensure it formats
                                // correctly in RTL
                                unicodeWrap(author),
                                pubDate.format(dateTimeFormat)
                            )
                        else -> authorTextView.visibility = View.GONE
                    }
                }
            }

            // Update state of notification toggle
            activity?.invalidateOptionsMenu()
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        arguments?.let { arguments ->
            _id = arguments.getLong(ARG_ID, ID_UNSET)
        }

        if (_id > ID_UNSET) {
            val itemId = _id
            val appContext = context?.applicationContext
            appContext?.let {
                lifecycleScope.launchWhenResumed {
                    viewModel.markAsReadAndNotified(_id)
                    cancelNotification(it, itemId)
                }
            }
        }

        setHasOptionsMenu(true)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        // Inflate the layout for this fragment
        val theLayout = if (TabletUtils.isTablet(activity)) {
            R.layout.fragment_reader_tablet
        } else {
            R.layout.fragment_reader
        }
        val rootView = inflater.inflate(theLayout, container, false)

        titleTextView = rootView.findViewById(R.id.story_title)
        bodyTextView = rootView.findViewById(R.id.story_body)
        authorTextView = rootView.findViewById(R.id.story_author)
        feedTitleTextView = rootView.findViewById(R.id.story_feedtitle)

        return rootView
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.reader, menu)

        // Set intent
        rssItemMenu?.let { rssItem ->
            // Show/Hide buttons
            menu.findItem(R.id.action_open_enclosure).isVisible = rssItem.enclosureLink != null
            menu.findItem(R.id.action_open_in_webview).isVisible = rssItem.link != null
            menu.findItem(R.id.action_open_in_browser).isVisible = rssItem.link != null
            menu.findItem(R.id.action_fetch_article).isVisible = !rssItem.fullTextByDefault
            // Add filename to tooltip
            if (rssItem.enclosureLink != null) {
                val filename = rssItem.enclosureFilename
                if (filename != null) {
                    menu.findItem(R.id.action_open_enclosure).title = filename
                }
            }
        }

        // Don't forget super call here
        super.onCreateOptionsMenu(menu, inflater)
    }

    enum class ReadingState {
        NOT_STARTED, PAUSED, READING
    }
    private var readingState: ReadingState = ReadingState.NOT_STARTED
    private val textToSpeechViewModel: TextToSpeechViewModel by instance()

    override fun onOptionsItemSelected(menuItem: MenuItem): Boolean {
        return when (menuItem.itemId) {
            R.id.action_fetch_article -> {
                lifecycleScope.launch {
                    liveText?.removeObservers(this@ReaderFragment)
                    liveText = viewModel.getLiveFullText(
                        getTextOptions(),
                        urlClickListener()
                    )
                    liveText?.observe(this@ReaderFragment) {
                        bodyTextView.text = it
                    }
                }
                true
            }
            R.id.action_open_in_webview -> {
                // Open in web view or custom tab
                lifecycleScope.launch {
                    context?.let { context ->
                        val rssItem = viewModel.getItem(_id)
                        rssItem.link?.let { link ->
                            when (prefs.openLinksWith) {
                                PREF_VAL_OPEN_WITH_CUSTOM_TAB -> {
                                    openLinkInCustomTab(context, link, _id)
                                }
                                else -> {
                                    findNavController().navigate(
                                        R.id.action_readerFragment_to_readerWebViewFragment,
                                        bundle {
                                            putString(ARG_URL, link)
                                            putString(ARG_ENCLOSURE, rssItem.enclosureLink)
                                            putLong(ARG_ID, _id)
                                        }
                                    )
                                }
                            }
                        }
                    }
                }
                true
            }
            R.id.action_open_in_browser -> {
                lifecycleScope.launch {
                    val link = viewModel.getItem(_id).link
                    if (link != null) {
                        context?.let { context ->
                            openLinkInBrowser(context, link)
                        }
                    }
                }
                true
            }
            R.id.action_open_enclosure -> {
                lifecycleScope.launch {
                    val link = viewModel.getItem(_id).enclosureLink
                    if (link != null) {
                        context?.let { context ->
                            openLinkInBrowser(context, link)
                        }
                    }
                }

                true
            }
            R.id.action_mark_as_unread -> {
                lifecycleScope.launch {
                    viewModel.markAsRead(_id, unread = true)
                }
                true
            }
            R.id.action_read_article_aloud -> {
                when (readingState) {
                    ReadingState.NOT_STARTED -> {
                        val fullText = liveText?.value.toString()
                        textToSpeechViewModel.textToSpeechClear()
                        textToSpeechViewModel.textToSpeechAddText(fullText)
                        textToSpeechViewModel.textToSpeechStart(lifecycleScope)
                        menuItem.title = getString(R.string.pause_reading)
                        readingState = ReadingState.READING
                    }
                    ReadingState.READING -> {
                        textToSpeechViewModel.textToSpeechPause()
                        menuItem.title = getString(R.string.resume_reading)
                        readingState = ReadingState.PAUSED
                    }
                    ReadingState.PAUSED -> {
                        textToSpeechViewModel.textToSpeechStart(lifecycleScope)
                        menuItem.title = getString(R.string.pause_reading)
                        readingState = ReadingState.READING
                    }
                }
                true
            }
            R.id.action_share -> {
                lifecycleScope.launch {
                    viewModel.getItem(_id).link?.let { link ->
                        val shareIntent = Intent(Intent.ACTION_SEND)
                        shareIntent.type = "text/plain"
                        shareIntent.putExtra(Intent.EXTRA_TEXT, link)

                        startActivity(Intent.createChooser(shareIntent, getString(R.string.share)))
                    }
                }
                true
            }
            else -> super.onOptionsItemSelected(menuItem)
        }
    }
}

fun Fragment.unicodeWrap(text: String): String =
    BidiFormatter.getInstance(getLocale()).unicodeWrap(text)

fun Fragment.getLocale(): Locale? =
    context?.getLocale()

fun Context.unicodeWrap(text: String): String =
    BidiFormatter.getInstance(getLocale()).unicodeWrap(text)

fun Context.getLocale(): Locale =
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
        resources.configuration.locales[0]
    } else {
        @Suppress("DEPRECATION")
        resources.configuration.locale
    }

fun Fragment.urlClickListener(): (link: String) -> Unit = { link ->
    context?.let { context ->
        val kodein: Kodein by closestKodein()
        val prefs: Prefs by kodein.instance()

        when (prefs.openLinksWith) {
            PREF_VAL_OPEN_WITH_CUSTOM_TAB -> {
                openLinkInCustomTab(context, link, null)
            }
            PREF_VAL_OPEN_WITH_WEBVIEW -> {
                findNavController().navigate(
                    R.id.action_readerFragment_to_readerWebViewFragment,
                    bundle {
                        putString(ARG_URL, link)
                    }
                )
            }
            else -> {
                openLinkInBrowser(context, link)
            }
        }
    }
}
